File: backend/app/api/v1/configuracion.py
from fastapi import APIRouter, Depends, File, UploadFile
from sqlalchemy.orm import Session

from app.api.deps import get_current_admin_full
from app.database.session import get_db
from app.schemas.configuracion import (
    ConfiguracionGeneralOut, ConfiguracionValor, PlantillaCorreoOut, VistaPreviaCorreoIn, VistaPreviaCorreoOut,
)
from app.schemas.smtp import SmtpConfigOut, SmtpConfigUpdate, SmtpPruebaRequest
from app.services.configuracion_service import ConfiguracionService
from app.services.smtp_config_service import SmtpConfigService

router = APIRouter(prefix="/configuracion", tags=["configuracion"])


@router.get("", response_model=ConfiguracionGeneralOut)
def obtener(db: Session = Depends(get_db)):
    """Pública: branding y datos no sensibles que necesita el portal."""
    return ConfiguracionService(db).obtener_publica()


@router.put("", status_code=204)
def actualizar(data: ConfiguracionValor, db: Session = Depends(get_db),
               admin: dict = Depends(get_current_admin_full)):
    ConfiguracionService(db).actualizar(data.clave, data.valor, admin["id"])


@router.get("/plantilla-correo", response_model=PlantillaCorreoOut)
def obtener_plantilla_correo(db: Session = Depends(get_db), _: dict = Depends(get_current_admin_full)):
    return ConfiguracionService(db).obtener_plantilla_correo()


@router.post("/plantilla-correo/vista-previa", response_model=VistaPreviaCorreoOut)
def previsualizar_plantilla_correo(data: VistaPreviaCorreoIn, db: Session = Depends(get_db),
                                    _: dict = Depends(get_current_admin_full)):
    return VistaPreviaCorreoOut(html=ConfiguracionService(db).previsualizar_plantilla_correo(data.cuerpo))


@router.post("/imagen-bienvenida")
async def subir_imagen_bienvenida(archivo: UploadFile = File(...), db: Session = Depends(get_db),
                                   admin: dict = Depends(get_current_admin_full)):
    url = await ConfiguracionService(db).guardar_imagen_bienvenida(archivo, admin["id"])
    return {"imagen_bienvenida_url": url}


@router.post("/imagen-correo")
async def subir_imagen_correo(archivo: UploadFile = File(...), db: Session = Depends(get_db),
                               admin: dict = Depends(get_current_admin_full)):
    url = await ConfiguracionService(db).guardar_imagen_correo(archivo, admin["id"])
    return {"email_confirmacion_imagen_url": url}


@router.get("/smtp", response_model=SmtpConfigOut)
def obtener_smtp(db: Session = Depends(get_db), _: dict = Depends(get_current_admin_full)):
    """La contraseña nunca se incluye en la respuesta, solo si hay una guardada."""
    return SmtpConfigService(db).obtener()


@router.put("/smtp", response_model=SmtpConfigOut)
def actualizar_smtp(data: SmtpConfigUpdate, db: Session = Depends(get_db),
                     admin: dict = Depends(get_current_admin_full)):
    return SmtpConfigService(db).actualizar(data, admin["id"])


@router.post("/smtp/prueba", status_code=204)
def probar_smtp(data: SmtpPruebaRequest, db: Session = Depends(get_db),
                 admin: dict = Depends(get_current_admin_full)):
    SmtpConfigService(db).enviar_prueba(data.destinatario, admin["id"])

File: frontend/src/services/configuracion.ts
import api from "./api";
import type { ConfiguracionGeneral, PlantillaCorreo, SmtpConfig, SmtpConfigUpdate } from "../types";

// Endpoint público (sin autenticación); se usa tanto desde el portal como desde el admin
// para precargar el formulario, por eso se deja en la instancia base.
export const obtenerConfiguracion = async () =>
  (await api.get<ConfiguracionGeneral>("/configuracion")).data;

export const actualizarConfiguracion = async (clave: string, valor: string | null) =>
  api.put("/configuracion", { clave, valor });

export const subirImagenBienvenida = async (archivo: File) => {
  const form = new FormData();
  form.append("archivo", archivo);
  return (await api.post<{ imagen_bienvenida_url: string }>("/configuracion/imagen-bienvenida", form))
    .data.imagen_bienvenida_url;
};

export const obtenerSemanaActiva = async () =>
  (await api.get<{ inicio: string | null; fin: string | null }>("/semana/activa")).data;

export const definirSemanaActiva = async (inicio: string, fin: string) =>
  api.put("/semana/activa", { inicio, fin });

export const reiniciarSemana = async (fecha_lunes: string) =>
  (await api.post<{ reservas_afectadas: number }>("/semana/reiniciar", null, {
    params: { fecha_lunes },
  })).data;

export const obtenerConfiguracionSmtp = async () =>
  (await api.get<SmtpConfig>("/configuracion/smtp")).data;

export const actualizarConfiguracionSmtp = async (payload: SmtpConfigUpdate) =>
  (await api.put<SmtpConfig>("/configuracion/smtp", payload)).data;

export const enviarCorreoPrueba = async (destinatario: string) =>
  api.post("/configuracion/smtp/prueba", { destinatario });

export const obtenerPlantillaCorreo = async () =>
  (await api.get<PlantillaCorreo>("/configuracion/plantilla-correo")).data;

export const previsualizarPlantillaCorreo = async (cuerpo: string) =>
  (await api.post<{ html: string }>("/configuracion/plantilla-correo/vista-previa", { cuerpo })).data.html;

export const subirImagenCorreo = async (archivo: File) => {
  const form = new FormData();
  form.append("archivo", archivo);
  return (await api.post<{ email_confirmacion_imagen_url: string }>("/configuracion/imagen-correo", form))
    .data.email_confirmacion_imagen_url;
};
